### ISSUE TITLE
[feature request] add Cubic Spline Distribution for samples and/or distribution "equalizer"

### ISSUE THREAD

**Issue by StableLlama** [bghira/SimpleTuner#2561]
To make it easy to use different sampling distributions, other than the standard sigmoid and the little more flexible Beta distribution, I suggest to add a Cubic Spline Distribution.

The user would configure an array of floats (JSON array or comma separated floats would work equally well) that define the probability at an equal spaced span between 0.0 and 1.0.

E.g.
- `[]` would result in the uniform distribution
- `[n0]` would result in the uniform distribution
- `[n0, n1]` would be a linear distribution with `scale*n0` at 0.0 and `scale*n1` at 1.0 with `scale` chosen so that it will be a distribution
- `[n0, n1, n2]` would be distribution with `scale*n0` at 0.0, `scale*n1` at 0.5 and `scale*n2` at 1.0 and a smooth transition between 0.0 and 0.5 and 0.5 and 1.0, with `scale` chosen so that it will be a distribution
- `[n0, n1, n2, n3]` would be distribution with `scale*n0` at 0.0, `scale*n1` at 0.33, `scale*n2` at 0.66 and `scale*n3` at 1.0 and a smooth transition between 0.0 and 0.33 and between 0.33 and 0.66 and between 0.66 and 1.0, with `scale` chosen so that it will be a distribution
- ...

A potential implementation (according to Gemini) for this distribution would be:
```python
import torch
import torch.nn.functional as F
from torch.distributions import Distribution, Uniform, constraints

class SmoothInterpolatedDistribution(Distribution):
    arg_constraints = {'weights': constraints.nonnegative}
    support = constraints.interval(0.0, 1.0)

    def __init__(self, weights, resolution=1024):
        """
        weights: Tensor of shape (K,) specifying density at K evenly spaced points.
        resolution: Internal grid size for normalization and sampling.
        """
        super().__init__(batch_shape=torch.Size(), event_shape=torch.Size())
        
        # Handle edge cases: 0 or 1 point -> Uniform
        if weights is None or weights.numel() <= 1:
            self.is_uniform = True
            self.weights = torch.tensor([1.0, 1.0])
        else:
            self.is_uniform = False
            self.weights = weights.float()

        self.K = self.weights.numel()
        self.res = resolution
        
        # 1. Create a high-res grid for the PDF
        grid_x = torch.linspace(0, 1, self.res)
        
        # 2. Smooth Interpolation
        # We use 'cubic' interpolation to get a smooth curve through the weights
        # Reshape to (Batch, Channel, Width) for F.interpolate
        w_input = self.weights.view(1, 1, -1)
        if self.K == 2:
            # Linear is the only smooth option for 2 points
            pdf_grid = F.interpolate(w_input, size=self.res, mode='linear', align_corners=True)
        else:
            # For 3+ points, we use cubic interpolation
            # Note: PyTorch interpolate 'area' or 'linear' are common, 
            # for 1D smooth cubic we can use a simple trick or 2D bicubic
            pdf_grid = F.interpolate(w_input.unsqueeze(-1), size=(self.res, 1), 
                                     mode='bicubic', align_corners=True).squeeze(-1)
        
        self.pdf_grid = pdf_grid.view(-1).clamp(min=1e-6) # Ensure non-negative
        
        # 3. Normalize to ensure integral = 1
        dx = 1.0 / (self.res - 1)
        total_area = torch.sum(self.pdf_grid) * dx
        self.pdf_grid = self.pdf_grid / total_area
        
        # 4. Compute CDF for sampling
        self.cdf_grid = torch.cumsum(self.pdf_grid * dx, dim=0)
        self.cdf_grid = self.cdf_grid / self.cdf_grid[-1] # Force last to 1.0
        self.grid_x = grid_x

    def sample(self, sample_shape=torch.Size()):
        # Inverse Transform Sampling
        u = torch.rand(sample_shape)
        # Find where u falls in the CDF grid
        indices = torch.searchsorted(self.cdf_grid, u)
        indices = indices.clamp(max=self.res - 1)
        return self.grid_x[indices]

    def log_prob(self, value):
        # Look up density at value (linear interpolation on the PDF grid)
        idx = (value * (self.res - 1)).long().clamp(0, self.res - 1)
        return torch.log(self.pdf_grid[idx])

# Examples of your requested cases:
# [] or [n0] -> Uniform
dist_uni = SmoothInterpolatedDistribution(torch.tensor([5.0])) 

# [n0, n1] -> Linear
dist_lin = SmoothInterpolatedDistribution(torch.tensor([0.1, 1.0]))

# [n0, n1, n2, n3] -> Smoothly interpolated peaks
dist_smooth = SmoothInterpolatedDistribution(torch.tensor([0.1, 2.0, 0.5, 3.0]))

samples = dist_smooth.sample((5000,))
```

**Comment by StableLlama** [comment]
As the generic sigmoid is working mostly quite well, and alternativ approach would be to continue and use it but other an "equalizer" where the user can adjust it slightly, i.e. give it a range where to increase the focus (i.e. sample more often) and others where it should concentrate less.

This can be done (thanks to Gemini; the idea is the sentence above, so when you don't want to read the AI output you can skip the rest of this comment):

In mathematical terms, what you are describing is **Probability Density Function (PDF) Reweighting** (or Modulation), followed by **Inverse Transform Sampling**.

Specifically:
1.  **Reweighting:** You are taking a base distribution $P_{base}(x)$ and a weighting function $w(x)$ (your equalizer) to create an unnormalized density $f(x) = P_{base}(x) \cdot w(x)$.
2.  **Normalization:** You compute the partition function $Z = \int f(x) dx$ to get the valid new distribution $P_{new}(x) = \frac{f(x)}{Z}$.
3.  **Inverse Transform Sampling:** To sample efficiently, you compute the Cumulative Distribution Function (CDF), invert it, and map uniform random numbers to your desired steps.

### Implementation Strategy

For high efficiency during training (where `sample` is called thousands of times), we should **not** use rejection sampling (which is slow and non-deterministic).

Instead, we will:
1.  **Discretize** the interval [0, 1] into a high-resolution buffer (e.g., 10,000 bins) during initialization.
2.  **Compute the PDF** by multiplying the base distribution (Sigmoid) by your interpolated equalizer array.
3.  **Compute the CDF** using a cumulative sum.
4.  **Sample** using `torch.searchsorted` (binary search) on the GPU. This makes sampling an $O(1)$ operation relative to the complexity of the distribution function.

### Python / PyTorch Implementation

Here is a robust, "drop-in" class. It handles the interpolation (linear), normalization, and efficient sampling.

```python
import torch
import torch.nn.functional as F

class LoRAStepEqualizer:
    def __init__(
        self, 
        equalizer_weights: list[float], 
        steps: int = 10000, 
        device: torch.device = None
    ):
        """
        Args:
            equalizer_weights: List of floats where index 0 is sigma 0.0, last is 1.0.
                               Values > 1.0 increase prob, < 1.0 decrease prob.
            steps: Resolution of the internal lookup table (higher = more precise).
            device: PyTorch device.
        """
        self.device = device if device else torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.steps = steps
        
        # 1. Validate Input
        if not equalizer_weights or len(equalizer_weights) < 2:
            print("Equalizer: Array too short, defaulting to uniform pass-through.")
            self.cdf = None
            return

        # 2. Prepare the Domain (0 to 1)
        # We create a linspace representing the sigma values from 0 to 1
        x = torch.linspace(0, 1, steps, device=self.device)

        # 3. Define Base Distribution (The "Sigmoid" distribution)
        # Note: Depending on your specific scheduler, 'sigmoid' might mean 
        # sampling t ~ Logistic or Normal. Here we model a standard logistic PDF 
        # centered at 0.5. If your base is Uniform, set base_pdf = 1.0.
        # This formula is the derivative of the sigmoid function (Standard Logistic Distribution).
        base_pdf = (torch.exp(x - 0.5) / (1 + torch.exp(x - 0.5))**2)
        
        # NOTE: If you simply want to modulate a Uniform distribution, 
        # uncomment the line below:
        # base_pdf = torch.ones_like(x) 

        # 4. Process the Equalizer Weights (Interpolation)
        weights_tensor = torch.tensor(equalizer_weights, dtype=torch.float32, device=self.device)
        
        # Reshape for F.interpolate: (Batch, Channel, Length) -> (1, 1, Length)
        weights_tensor = weights_tensor.view(1, 1, -1)
        
        # Interpolate weights to match the resolution of our 'steps' (Linear interpolation)
        # align_corners=True ensures index 0 maps to 0.0 and last index maps to 1.0 exactly
        interpolated_weights = F.interpolate(
            weights_tensor, 
            size=steps, 
            mode='linear', 
            align_corners=True
        ).view(-1)

        # 5. Apply the Equalizer (Reweighting)
        # New Unnormalized Density = Base PDF * User Weights
        weighted_pdf = base_pdf * interpolated_weights

        # 6. Normalize to create a valid Probability Distribution
        # We divide by the sum to ensure the area under the curve equals 1
        pdf_sum = weighted_pdf.sum()
        if pdf_sum == 0:
             # Fallback if weights kill all probability
            prob_dist = torch.ones_like(x) / steps
        else:
            prob_dist = weighted_pdf / pdf_sum

        # 7. Create Cumulative Distribution Function (CDF) for Inverse Sampling
        self.cdf = torch.cumsum(prob_dist, dim=0)
        
        # Ensure the last value is exactly 1.0 to avoid numerical issues
        self.cdf[-1] = 1.0

    def sample(self, batch_size: int = 1) -> torch.Tensor:
        """
        Returns a tensor of shape (batch_size,) with values in range [0.0, 1.0]
        distributed according to the equalized curve.
        """
        # If initialization failed or skipped, return standard Uniform samples
        if self.cdf is None:
            return torch.rand(batch_size, device=self.device)

        # 1. Generate Uniform Random Numbers [0, 1]
        r = torch.rand(batch_size, device=self.device)

        # 2. Inverse Transform Sampling using binary search (searchsorted)
        # Find the indices where the random values would fit in the CDF
        indices = torch.searchsorted(self.cdf, r)

        # 3. Map indices back to float range [0.0, 1.0]
        sampled_sigmas = indices.float() / self.steps
        
        return torch.clamp(sampled_sigmas, 0.0, 1.0)

# --- Example Usage ---

# Example: We want to focus heavily on the middle steps and ignore the very start/end.
# 0.0 -> 0.1x prob
# 0.5 -> 3.0x prob (Focus hard here)
# 1.0 -> 0.1x prob
user_weights = [0.1, 1.0, 3.0, 1.0, 0.1] 

# Initialize
equalizer = LoRAStepEqualizer(user_weights, steps=10000)

# Sample a batch of steps for training
step_values = equalizer.sample(batch_size=5)

print(f"Sampled Steps: {step_values}")
# Expected: Values will cluster around 0.5, very few near 0.0 or 1.0.
```

### Why this works well for LoRA training

1.  **Interpolation**: By using `F.interpolate` with `mode='linear'`, we gracefully handle your requirement that the user can pass `[0.5, 2.0]` (a slope) or `[1, 5, 1]` (a bell curve) or any arbitrary complexity, and it smoothly maps to the sampling domain.
2.  **SearchSorted**: This is the key to efficiency. Instead of complex math during every training step, we do a simple binary search on a pre-sorted array (the CDF). This is natively optimized in CUDA.
3.  **Stability**: By normalizing the PDF (`prob_dist = weighted_pdf / pdf_sum`), we guarantee mathematically valid probabilities regardless of how high or low the user sets the multiplier numbers.

### A Note on the "Base Distribution"

In the code above (Step 3), I assumed your "sigmoid distribution" implied a **Logistic Distribution** (bell curve) for the timesteps.
*   If your original code simply samples uniformly `torch.rand()` and then applies a sigmoid *function* to those values later to get sigmas, you should change `base_pdf` to `torch.ones_like(x)` (Uniform).
*   The "Equalizer" will then act as the sole shaper of the probability curve.
